driver-core/src/main/com/mongodb/operation/OperationHelper.java

```diff
@@ -444,6 +526,76 @@ static void withConnection(final AsyncReadBinding binding, final AsyncCallableWi
         binding.getReadConnectionSource(errorHandlingCallback(new AsyncCallableWithConnectionAndSourceCallback(callable), LOGGER));
     }
 
+    static <T> T retryableWithConnection(final WriteBinding binding, final boolean retryable,
+                                         final CallableWithConnectionAndTransactionNumber<T> callable) {
+        final Long txnNumber = retryable ? binding.getSessionContext().advanceTransactionNumber() : null;
+        return withConnection(binding, new CallableWithConnection<T>() {
+            @Override
+            public T call(final Connection connection) {
+                try {
+                    return callable.call(connection, txnNumber);
+                } catch (MongoException e) {
+                    if (!retryable) {
+                        throw e;
+                    } else {
+                        final MongoException originalException = e;
+                        return withConnection(binding, new CallableWithConnection<T>() {
+                            @Override
+                            public T call(final Connection connection) {
+                                if (!serverIsAtLeastVersionThreeDotSix(connection.getDescription())) {
+                                    throw originalException;
+                                }
+                                return callable.call(connection, txnNumber);
+                            }
+                        });
+                    }
+                }
+            }
+        });
+    }
+
+    static <T> void retryableWithConnection(final AsyncWriteBinding binding, final boolean retryable,
+                                            final SingleResultCallback<T> callback,
+                                            final AsyncCallableWithConnectionAndTransactionNumber<T> callable) {
+        final SingleResultCallback<T> errHandlingCallback = errorHandlingCallback(callback, LOGGER);
+        final Long txnNumber = retryable ? binding.getSessionContext().advanceTransactionNumber() : null;
+        withConnection(binding, new AsyncCallableWithConnection() {
+            @Override
+            public void call(final AsyncConnection connection, final Throwable t) {
+                if (t != null) {
+                    errHandlingCallback.onResult(null, t);
+                } else {
+                    callable.call(connection, txnNumber, new SingleResultCallback<T>() {
+                        @Override
+                        public void onResult(final T result, final Throwable originalError) {
+                            if (originalError != null) {
+                                if (!retryable || !(originalError instanceof MongoException)) {
+                                    releasingCallback(errHandlingCallback, connection).onResult(null, originalError);
+                                } else {
+                                    connection.release();
```

Comment:
Looks good in the async case: the first connection is released prior to the second connection being checkout out.